### ISSUE TITLE
`bundle lock --add-platform` is recommended, but fails with error

### ISSUE THREAD

**Issue by jrochkind** [bundler/bundler#6655]
Complete repro Gemfile:

~~~ruby
source 'https://rubygems.org'

group :debug do
  gem "ruby-debug", :platform => "jruby"
  gem "byebug", :platform => "mri"
end
~~~

Pretty reasonable, platform-appropriate debuggers, right?

Running in jruby 9.2.0.0, with bundler 1.16.3. 

```bash
$ bundle install
The dependency byebug (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for java but the dependency is only for ruby. To add those platforms to the bundle, run `bundle lock --add-platform ruby`.
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...
Using bundler 1.16.3
Using columnize 0.9.0
Using linecache 1.3.1 (java)
Using ruby-debug-base 0.10.6 (java)
Using ruby-debug 0.10.6
Bundle complete! 2 Gemfile dependencies, 5 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
```

Okay, it's complaining that I have an MRI-specific gem in there but am on JRuby... not totally sure why it's complaining, this seems to suggest that platform-specific deps will _always_ result in this complaint/warning, since the only point of them is to be _sometimes_ not installed?

But okay, it's suggesting something I should do. `bundle lock --add-platform ruby`  Okay, let's try. 

```bash
$ bundle lock --add-platform ruby
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...
Bundler could not find compatible versions for gem "ruby-debug-base":
  In snapshot (Gemfile.lock):
    ruby-debug-base (= 0.10.6)

  In Gemfile:
    ruby-debug java was resolved to 0.10.6, which depends on
      ruby-debug-base (~> 0.10.6.0)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
```

I have no idea what this means or what I'm supposed to do about it. Is it a bug that this error message is happening at all?  I think so maybe?  It's like... you can't do `bundle lock --add-platform ruby` if you are in JRuby and have some JRuby-specific gems... but that seems to make `bundle lock --add-platform` kind of not so useful, if that's what's going on?  I don't entirely understand what `bundle lock --add-platform` is supposed to do though. 

Running `bundle update` doesn't do anything useful, just repeats the same warning message from the original `bundle install` above.

**Comment by segiddins** [comment]
This looks like a bug, thanks for the Gemfile that reproduces it!

**Comment by segiddins** [comment]
So looking into this, it looks like the issue is that https://rubygems.org/gems/ruby-debug/versions/0.10.6 is a "ruby" platform gem, but its dependency 
https://rubygems.org/gems/ruby-debug-base only has a "java" platform version 0.10.6. This causes issues with resolution, since bundler does not use the `platform` option in the Gemfile when reading in a lockfile, only the platform of the gem in the lockfile. And since Bundler sees a `ruby` platform `ruby-debug`, it thinks it needs to resolve `ruby-debug` (and its dependencies) for the ruby platform.

**Comment by colby-swandale** [comment]
@segiddins is there any actions we can do to resolve this problem?

**Comment by segiddins** [comment]
Maybe? Nothing simple tho, it would require a rethink of Bundlers approach to platsfoems and resolution

**Comment by jrochkind** [comment]
Hmm. I'm confused about if `--add-platform` will _always_ result in this error, or just under certain cirucmstances?  Does `--add-platform` work at all?

When bundler says:

> Bundler is installing for java but the dependency is only for ruby. To add those platforms to the bundle, run `bundle lock --add-platform ruby`.

Will this advice _ever_ work?  In what circumstances is it intended? In what circumstances does it work?  (Did it _used_ to work more than it does now, is it a regression?)

**Comment by jrochkind** [comment]
Note that you get this warning on `rails new` with rails 5.2. 

```
The dependency tzinfo-data (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for x86-mingw32, x86-mswin32, x64-mingw32, java. To add those platforms to the bundle, run `bundle lock --add-platform x86-mingw32 x86-mswin32 x64-mingw32 java`
```

However the recommended `bundle lock` command _does_ work in this case.

**Comment by segiddins** [comment]
> Does --add-platform work at all

It certainly does work in the majority of situations it just doesnt work in this _particular_ one due to a ruby platform gem that can't actually be installed on the ruby platform

**Comment by segiddins** [comment]
> Is the problem is that ruby-debug has buggy dependencies of some kind?

Yeah, sort of. It's just that it _says_ it should be installable on _any_ Ruby, but in fact its dependencies will only ever resolve for JRuby
